File: media_bot.py
import os
import re
import asyncio
import logging
from dotenv import load_dotenv
from pyrogram import Client, enums
import yt_dlp

# Load environment variables
load_dotenv()

# Configuration
API_ID = os.getenv('API_ID')
API_HASH = os.getenv('API_HASH')
BOT_TOKEN = os.getenv('BOT_TOKEN')
CHANNEL_ID = os.getenv('TELEGRAM_CHANNEL_ID')

# Ensure configuration exists
if not all([API_ID, API_HASH, BOT_TOKEN, CHANNEL_ID]):
    print("Error: Missing configuration in .env file.")
    exit(1)

# Logging setup
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Initialize Pyrogram Client
# Using in-memory session to ensure it acts as a Bot and not a User
app = Client(
    "media_bot_session",
    api_id=API_ID,
    api_hash=API_HASH,
    bot_token=BOT_TOKEN,
    in_memory=True
)

def get_channel_id(channel_id_str):
    """
    Resolves the channel ID.
    Pyrogram handles strings (@channel) and integers (-100...) natively,
    but we ensure integers are cast correctly.
    """
    if channel_id_str.lstrip('-').isdigit():
        return int(channel_id_str)
    return channel_id_str

async def download_media(url):
    """Downloads media from the given URL using yt-dlp."""
    output_dir = 'downloads'
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    # Use a specific template to control filenames
    output_template = os.path.join(output_dir, '%(id)s.%(ext)s')

    ydl_opts = {
        'outtmpl': output_template,
        'quiet': True,
        'no_warnings': True,
        'max_filesize': 2000 * 1024 * 1024, # 2GB
        # 'noplaylist': True, # Default behavior
    }

    try:
        logger.info(f"Downloading: {url}")
        # Run yt-dlp in a separate thread to avoid blocking the async loop
        def run_yt_dlp():
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(url, download=True)
                return info

        loop = asyncio.get_event_loop()
        # Using executor for blocking call
        await loop.run_in_executor(None, run_yt_dlp)

        # Return list of files in output_dir
        # Since we clean up, any file here is new
        return [os.path.join(output_dir, f) for f in os.listdir(output_dir)]

    except Exception as e:
        logger.error(f"Download error for {url}: {e}")
        return []

async def process_messages():
    """Poller to process messages."""

    channel_target = get_channel_id(CHANNEL_ID)
    logger.info(f"Monitoring channel: {CHANNEL_ID}")

    async with app:
        while True:
            try:
                # Fetch history (last 10 messages)
                # Pyrogram's get_chat_history is async generator
                history = app.get_chat_history(channel_target, limit=10)

                async for message in history:
                    if not message.text:
                        continue

                    # Check criteria: "New Saved Posts:" and not "✅"
                    if "New Saved Posts:" in message.text and "✅" not in message.text:
                        logger.info(f"Processing message {message.id}...")

                        # Extract URLs
                        # Regex modified to handle Markdown links [Title](URL) or plain URL
                        # Captures URL inside () or standalone
                        url_regex = r"https?://(?:www\.)?(?:reddit\.com|redd\.it)/[^\s\)]+"
                        urls = re.findall(url_regex, message.text)

                        if not urls:
                            continue

                        files_to_upload = []

                        for url in urls:
                            files = await download_media(url)
                            files_to_upload.extend(files)

                        # Upload
                        uploaded_count = 0
                        for file_path in files_to_upload:
                            if os.path.exists(file_path):
                                logger.info(f"Uploading {file_path}...")
                                try:
                                    # Pyrogram supports uploading files directly
                                    # We try to determine type, or just send_document/video
                                    # Reddit media is usually video or image.
                                    # send_document is safest for "file", send_video/photo for stream.
                                    # User mentioned "stream it". send_video/send_photo renders it in stream.
                                    ext = os.path.splitext(file_path)[1].lower()
                                    if ext in ['.jpg', '.jpeg', '.png', '.webp']:
                                        await app.send_photo(channel_target, file_path)
                                    elif ext in ['.mp4', '.mkv', '.webm', '.gif']:
                                        await app.send_video(channel_target, file_path)
                                    else:
                                        await app.send_document(channel_target, file_path)

                                    uploaded_count += 1
                                except Exception as err:
                                    logger.error(f"Upload failed for {file_path}: {err}")
                                finally:
                                    # Delete immediately after upload attempt
                                    if os.path.exists(file_path):
                                        os.remove(file_path)

                        # Mark as processed
                        # We append to the message.
                        new_text = message.text + "\n\n✅ Processed & Downloaded"
                        try:
                            # Keep the formatting of the original message (Markdown)
                            await app.edit_message_text(channel_target, message.id, new_text, disable_web_page_preview=True)
                            logger.info("Message marked as processed.")
                        except Exception as err:
                            logger.error(f"Failed to edit message: {err}")

                        # Final Cleanup of downloads dir (if any leftovers)
                        output_dir = 'downloads'
                        if os.path.exists(output_dir):
                            for f in os.listdir(output_dir):
                                p = os.path.join(output_dir, f)
                                if os.path.exists(p):
                                    os.remove(p)
                            os.rmdir(output_dir)

            except Exception as e:
                logger.error(f"Error in loop: {e}")
                # Wait a bit before retrying if error
                await asyncio.sleep(5)

            # Sleep 15s before next poll
            await asyncio.sleep(15)

if __name__ == '__main__':
    # Run the poller
    try:
        asyncio.run(process_messages())
    except KeyboardInterrupt:
        print("Stopped.")

File: media_bot.js
require('dotenv').config();
const { TelegramClient } = require('telegram');
const { StringSession } = require('telegram/sessions');
const { Api } = require('telegram/tl');
const input = require('input'); // npm install input
const fs = require('fs');
const path = require('path');
const YTDlpWrap = require('yt-dlp-wrap').default;

// Configuration
const API_ID = Number(process.env.API_ID);
const API_HASH = process.env.API_HASH;
const BOT_TOKEN = process.env.BOT_TOKEN;
const CHANNEL_ID = process.env.TELEGRAM_CHANNEL_ID;

// Session management
const SESSION_FILE = 'session.txt';
let stringSession = new StringSession('');

if (fs.existsSync(SESSION_FILE)) {
    stringSession = new StringSession(fs.readFileSync(SESSION_FILE, 'utf-8'));
}

// Initialize yt-dlp
const ytDlpBinaryPath = path.join(__dirname, 'yt-dlp');
const ytDlpWrap = new YTDlpWrap(ytDlpBinaryPath);

async function ensureYtDlp() {
    if (!fs.existsSync(ytDlpBinaryPath)) {
        console.log('Downloading yt-dlp binary...');
        await YTDlpWrap.downloadFromGithub(ytDlpBinaryPath);
        console.log('Downloaded yt-dlp.');
    }
}

async function downloadMedia(url) {
    const outputDir = path.join(__dirname, 'downloads');
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir);
    }

    // Output template: downloads/id.ext
    const outputTemplate = path.join(outputDir, '%(id)s.%(ext)s');

    try {
        console.log(`Downloading: ${url}`);
        // Get metadata first to check for playlist/entries
        const metadata = await ytDlpWrap.execPromise([
            url,
            '--dump-json',
            '--no-playlist' // We treat playlists as single entries or iterate?
                            // Python script handled entries. Let's do simple download.
        ]);

        // Actually download
        // We use execPromise to run the command
        await ytDlpWrap.execPromise([
            url,
            '-o', outputTemplate,
            '--no-warnings',
            '--max-filesize', '2000m'
        ]);

        // Find downloaded files
        // Since we don't know the exact extension, we list the dir and match ID?
        // Or we parse the JSON dump.
        // Let's rely on listing the directory for files that were just modified or match the pattern?
        // Simpler: List all files in downloads/ and upload them?
        // Or better: yt-dlp prints the filename.
        // Let's use a simpler approach: list files in downloads dir.

        const files = fs.readdirSync(outputDir).map(f => path.join(outputDir, f));
        return files;
    } catch (e) {
        console.error('Download error:', e.message);
        return [];
    }
}

(async () => {
    await ensureYtDlp();

    console.log("Starting Telegram Bot...");
    const client = new TelegramClient(stringSession, API_ID, API_HASH, {
        connectionRetries: 5,
    });

    await client.start({
        botAuthToken: BOT_TOKEN,
    });

    // Save session
    fs.writeFileSync(SESSION_FILE, client.session.save());

    console.log("Bot connected.");

    // Resolve Channel ID (handle @username or -100 ID)
    let entity;
    try {
        // If it looks like an int, parse it.
        const id = isNaN(CHANNEL_ID) ? CHANNEL_ID : BigInt(CHANNEL_ID);
        entity = await client.getEntity(id);
    } catch (e) {
        console.error("Could not find channel:", e);
        return;
    }

    console.log(`Monitoring ${entity.title || CHANNEL_ID}...`);

    while (true) {
        try {
            // Get last 10 messages
            const messages = await client.getMessages(entity, { limit: 10 });

            for (const message of messages) {
                if (!message.text) continue;

                // Check criteria: "New Saved Posts" and not "✅"
                if (message.text.includes("New Saved Posts:") && !message.text.includes("✅")) {
                    console.log(`Processing message ${message.id}...`);

                    // Extract URLs
                    const urlRegex = /(https?:\/\/(?:www\.)?(?:reddit\.com|redd\.it)\/[^\s]+)/g;
                    const urls = message.text.match(urlRegex);

                    if (!urls) continue;

                    const filesToUpload = [];

                    for (const url of urls) {
                        const files = await downloadMedia(url);
                        filesToUpload.push(...files);
                    }

                    // Upload
                    for (const file of filesToUpload) {
                        if (fs.existsSync(file)) {
                            console.log(`Uploading ${file}...`);
                            try {
                                await client.sendFile(entity, {
                                    file: file,
                                });
                                fs.unlinkSync(file); // Delete after upload
                            } catch (err) {
                                console.error("Upload failed:", err);
                            }
                        }
                    }

                    // Mark as processed
                    // GramJS editMessage
                    const newText = message.text + "\n\n✅ Processed & Downloaded";
                    try {
                        await client.editMessage(entity, {
                            message: message.id,
                            text: newText,
                            linkPreview: false
                        });
                        console.log("Message marked as processed.");
                    } catch (err) {
                        console.error("Failed to edit message:", err);
                    }

                    // Cleanup downloads dir
                    const downloadDir = path.join(__dirname, 'downloads');
                    if (fs.existsSync(downloadDir)) {
                         const remaining = fs.readdirSync(downloadDir);
                         if (remaining.length === 0) fs.rmdirSync(downloadDir);
                    }
                }
            }
        } catch (e) {
            console.error("Error in loop:", e);
        }

        // Sleep 15s
        await new Promise(r => setTimeout(r, 15000));
    }
})();
